sql/rpl_master_info_file.hh

```diff
@@ -0,0 +1,622 @@
+      {
+        auto default_use_gtid=
+          static_cast<enum_master_use_gtid>(::master_use_gtid);
+        return default_use_gtid >= enum_master_use_gtid::DEFAULT ? (
+          gtid_supported ?
+            enum_master_use_gtid::SLAVE_POS : enum_master_use_gtid::NO
+        ) : default_use_gtid;
+      }
+      return mode;
+    }
+    operator bool()
+    { return operator enum_master_use_gtid() != enum_master_use_gtid::NO; }
+    auto &operator=(enum_master_use_gtid mode)
+    {
+      this->mode= mode;
+      DBUG_ASSERT(!is_default());
+      return *this;
+    }
+    bool is_default() override
+    { return mode >= enum_master_use_gtid::DEFAULT; }
+    bool set_default() override
+    {
+      mode= enum_master_use_gtid::DEFAULT;
+      gtid_supported= true;
+      return false;
+    }
+    /** @return
+      `true` if the line is a @ref enum_master_use_gtid,
+      `false` otherwise or on error
+    */
+    bool load_from(IO_CACHE *file) override
+    {
+      /**
+        Only 3 chars are required for the enum,
+        similar to @ref OptionalBoolField::load_from()
+      */
+      char buf[3];
+      if (!my_b_gets(file, buf, 3) ||
+          buf[1] != '\n' ||
+          buf[0] > /* SLAVE_POS */ '2' || buf[0] < /* NO */ '0')
+        return true;
+      operator=(static_cast<enum_master_use_gtid>(buf[0] - '0'));
+      return false;
+    }
+    void save_to(IO_CACHE *file) override
+    {
+      my_b_write_byte(file,
+        '0' + static_cast<unsigned char>(operator enum_master_use_gtid()));
+    }
+  }
+  /// Using_Gtid
+  master_use_gtid;
+
+  /// Replicate_Do_Domain_Ids
+  IDArrayField do_domain_ids;
+  /// Replicate_Ignore_Domain_Ids
+  IDArrayField ignore_domain_ids;
+  OptionalIntField<::master_retry_count> master_retry_count;
+
+  /**
+    This is a non-negative `DECIMAL(10,3)` seconds field internally
+    calculated as an unsigned integer milliseconds field.
+    It has a `DEFAULT` value of @ref ::master_heartbeat_period,
+    which in turn has a `DEFAULT` value of `@@slave_net_timeout / 2` seconds.
+  */
+  struct: OptionalField<uint32_t>
+  {
+    using OptionalField::operator=;
+    operator uint32_t() override
+    {
+      return is_default() ? ::master_heartbeat_period.value_or(
+        MY_MIN(slave_net_timeout*500ULL, SLAVE_MAX_HEARTBEAT_PERIOD)
+      ) : *(OptionalField<uint32_t>::optional);
+    }
+    bool load_from(IO_CACHE *file) override
+    {
+      /// Read in floating point first to validate the range
+      double seconds;
+      /**
+        Number of chars OptionalIntField::load_from() uses plus
+        1 for the decimal point; truncate the excess precision,
+        which there should not be unless the file is edited externally.
+      */
+      char buf[IntIOCache::BUF_SIZE<uint32_t> + 3];
+      size_t length= my_b_gets(file, buf, sizeof(buf));
+      if (!length)
+        return true;
+    #ifdef __clang__
+      // FIXME: floating-point variants of `std::from_chars()` not supported
```

Comment:
According to [this Stack Overflow answer](https://stackoverflow.com/a/78471596), `libc++` (Clang) doesn’t support floating-point variants of `std::from_chars()`.
How did it pass our (relatively recent) C++17 requirement, then?

Reply:
LLVM Clang supports `std::from_chars()` for `float` and `double` since [v20](https://releases.llvm.org/20.1.0/projects/libcxx/docs/Status/Cxx17.html) (llvm/llvm-project#91651).
No word on Apple Clang. 🍎